Implement T-Money balance parsing

justuswilhelm opened the pull request:
I've written a small plugin for Metroflip that lets you read the balance off a T-Money card. I've tested it with my T-Money card bought in 2018 in Seoul, Korea.

I wasn't sure whether adding my code to `metroflip_scene_detect_scan_callback` is a good pattern. Metroflip might recognize other cards as T-Money that support the same Iso14443_4a protocol. Let me know if there's a better way.

The ISO7816 and KSX6924 reading mechanism is inspired by metrodroid.

luu176 committed:
fix renfe issue

justuswilhelm committed:
Parse T-Money balance

This copies the structure from calypso.c, mostly
The ISO7816 and KSX6924 reading mechanism is inspired by metrodroid

luu176 commented:
Hi @justuswilhelm, 

Thanks for your contribution! Wondering how we're going to implement this as mifare classic runs on that same protocol, it would probablt be best to check iso 14a at the very bottom of the else if loop. to make sure no other cards detect as T-Money. It would also be good for you to find something unique about T-Money cards, so if someone scans a regular 14a card that isn;t supported it also doesnt get detected as T-Money.

justuswilhelm commented:
Thank you for your feedback. I'll take a look at this and find out if there's a way I can use the Mifare part for this somehow. I'll get back some time this month, so please wait.

justuswilhelm commented:
> Hi @justuswilhelm,
> 
> Thanks for your contribution! Wondering how we're going to implement this as mifare classic runs on that same protocol, it would probablt be best to check iso 14a at the very bottom of the else if loop. to make sure no other cards detect as T-Money. It would also be good for you to find something unique about T-Money cards, so if someone scans a regular 14a card that isn;t supported it also doesnt get detected as T-Money.

Hi, I haven't had time to look into this further this month, unfortunately. For the first solution, putting the check at the bottom, I can implement this. On the other hand, to check "unique" properties of the T-Money card would possibly require creating a Poller instance during the general card detection loop.

Do you have any suggestions for how to implement this?

luu176 commented:
Hi, I am currently working on a detection scheme for knowing what type of card it is based on the ATR (Answer To Reset). This value is unique to each public transport, so if you can find the ATR for T-Money or maybe I can ask around, and then I will implement.

luu176 committed:
Merge branch 'dev' into t-money

luu176 commented:
I found this ATR:

3B 88 80 01 04 02 00 20 00 71 C1 40 DF


can you verify the bytes in that payload that DO match the ATR in your card please?

luu176 commented:
Can you read the card in the main nfc application in flipper and go to more>Info then scroll all the way down to historical bytes and it write that in here

justuswilhelm commented:
Here's everything from the Flipper Zero dump in the `.nfc` file:

```
Filetype: Flipper NFC device
Version: 4
# Device type can be ISO14443-3A, ISO14443-3B, ISO14443-4A, ISO14443-4B, ISO15693-3, FeliCa, NTAG/Ultralight, Mifare Classic, Mifare Plus, Mifare DESFire, SLIX, ST25TB, NTAG4xx, Type 4 Tag, EMV
Device type: ISO14443-4A
# UID is common for all formats
UID: XX XX XX XX
# ISO14443-3A specific data
ATQA: 00 04
SAK: 20
# ISO14443-4A specific data
T0: 78
TA(1): 80
TB(1): B0
TC(1): 02
T1...Tk: 04 09 22 02
```

luu176 commented:
thanks, i've added your ATR, then I will fix your PR and merge

luu176 committed:
Refactor key verification and remove unused functions

Removed two_cities_verify and renfe_regular_verify functions, updated key values for troika_4k_key and renfe_suma10_1k_keys.

luu176 committed:
Update verification keys and add two cities support

luu176 committed:
Update NFC protocol handling in metroflip_scene

justuswilhelm commented:
Thank you!
